# 实验二


## 0.数据导入

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from pylab import rcParams
rcParams['font.sans-serif'] = ['SimHei']
rcParams['axes.unicode_minus'] = False

In [13]:
data = pd.read_excel('1 第4章 数据特征的描述.xlsx',None)
data_mean = data['算术平均数']
data_hmean = data['调和平均数']
data_w_hmean = data['加权调和平均数']
data

{'调和平均数':     单价  购买金额（元）
 0  2.0        1
 1  2.3        1
 2  2.6        1,
 '算术平均数':      学生编号  月消费支出额（元）
 0       1        410
 1       2        350
 2       3        466
 3       4        385
 4       5        500
 ..    ...        ...
 105   106        811
 106   107       1109
 107   108       1276
 108   109        745
 109   110        678
 
 [110 rows x 2 columns],
 '加权调和平均数':     单价  购买金额（元）
 0  2.0      1.0
 1  2.3      1.2
 2  2.6      1.5}

## 1.第一题

### 1.1算术平均数

In [14]:
arithmetic_mean = data_mean['月消费支出额（元）'].mean()
arithmetic_mean

597.0181818181818

### 1.2调和平均数

In [15]:
harmonic_mean = stats.hmean(data_hmean['单价'])
harmonic_mean

2.2737642585551328

### 1.3中位数

In [16]:
median = data_mean['月消费支出额（元）'].median()
median

562.0

### 1.4众数

In [17]:
mode = data_mean['月消费支出额（元）'].mode()
mode[0] # mode是dataframe形式

520

### 1.5几何平均数

In [18]:
geometric_mean = stats.gmean(data_mean['月消费支出额（元）'])
geometric_mean

579.0721701196758

### 1.6加权调和平均数

In [19]:
x = data_w_hmean['购买金额（元）'] / data_w_hmean['单价']
x_sum = x.sum()
weighted_harmonic_mean = data_w_hmean['购买金额（元）'].sum() / x_sum
weighted_harmonic_mean

2.3144351464435147

## 2.第二题

### 2.1最大值

In [20]:
data_mean['月消费支出额（元）'].max()

1276

### 2.2最小值

In [21]:
data_mean['月消费支出额（元）'].min()

347

### 2.3四分位点

In [22]:
data_mean['月消费支出额（元）'].quantile([0.25,0.5,0.75])

0.25    496.25
0.50    562.00
0.75    675.50
Name: 月消费支出额（元）, dtype: float64

### 2.4样本标准差

In [23]:
data_mean['月消费支出额（元）'].std()

156.96291748229515

### 2.5峰度系数

In [24]:
data_mean['月消费支出额（元）'].kurt()

3.0365767586677097

### 2.6偏态系数

In [25]:
data_mean['月消费支出额（元）'].skew()

1.3732950318171815

## 3.第三题

In [26]:
food_data = pd.read_excel('2 descriptive stats.xlsx')
food_data

Date  Cakes  Pies  Cookies  Smoothies  Coffee
0    2013-01-01     79    46      518         60     233
1    2013-01-02     91    50      539        161     427
2    2013-01-03     47    60      222        166     347
3    2013-01-04     89    64      734        153     358
4    2013-01-05    112    73      764        240     392
...         ...    ...   ...      ...        ...     ...
1090 2015-12-27    136    49      700        265     483
1091 2015-12-28     77    45      605         69     408
1092 2015-12-29     93    56      508        185     452
1093 2015-12-30     62    28      458        141     356
1094 2015-12-31     65    56      491        190     446

[1095 rows x 6 columns]

### 偏态系数

In [27]:
food_data['Cookies'].skew()

0.4402865556120628

### 算术平均数、众数、中位数

In [28]:
mean_cookies = food_data['Cookies'].mean()
mode_cookies = food_data['Cookies'].mode()
median_cookies = food_data['Cookies'].median()
print("算术平均数为：",mean_cookies)
print("众数为：",mode_cookies)
print("中位数为：",median_cookies)

算术平均数为： 540.4511415525114
众数为： 0    397
1    552
dtype: int64
中位数为： 533.0


注意到有两个众数：<br>
算术平均数(540) >   中位数(533)      >   众数(397) <br>
众数(552)      >   算术平均数(540)   >   中位数(533) <br>
<br>
注意到：
偏态系数 > 0 且 其中一个众数满足 $\bar x > M_o > M_e $ <br>
所以得出结论：
Cookies的日销量是偏斜的

## 4.第四题
根据经验法则，你预计在95％的日子中，每天对沙冰（Smoothies）的需求位于___和 ___之间。确定沙冰需求是异常点的天数比例。

In [29]:
mean_Smoothies = food_data['Smoothies'].mean()
std_Smoothies = food_data['Smoothies'].std()
print("算术平均数为：",mean_Smoothies)
print("标准差为：",std_Smoothies)

算术平均数为： 218.46666666666667
标准差为： 74.35093016617384


In [30]:
left = mean_Smoothies - 2 * std_Smoothies
right = mean_Smoothies + 2 * std_Smoothies
print("在95％的日子中，每天对沙冰（Smoothies）的需求位于{}和{}之间".format(round(left),round(right)))

在95％的日子中，每天对沙冰（Smoothies）的需求位于70和367之间


### 异常点

In [31]:
abnormal_left = mean_Smoothies - 3 * std_Smoothies
abnormal_right = mean_Smoothies + 3 * std_Smoothies
print("在{}和{}之外的点均为异常点".format(round(abnormal_left),round(abnormal_right)))

在-5和442之外的点均为异常点


In [32]:
cnt = 0
for i in food_data['Smoothies'].values:
    if i < abnormal_left or i > abnormal_right:
        cnt += 1
print("异常点数量：{}".format(cnt))
print("异常点天数所占比例为{:.3f}%".format(cnt / len(food_data['Smoothies']) * 100))

异常点数量：8
异常点天数所占比例为0.731%


## 5.第五题
在每天至少售出500个饼干（cookies）的日子中找出蛋糕销售的中位数

In [33]:
food_data[food_data['Cookies'] >= 500]['Cookies'].median()

600.0

在每天至少售出500个饼干（cookies）的日子中蛋糕销售的中位数为 600（个）